pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java

```diff
@@ -30,9 +30,35 @@
   /**
    * Return whether the client has data access to the given table.
    *
+   * Note: This method is only used for read access. It will be deprecated soon and its usage will be replaced by
+   * `hasAccess` method with AccessType.READ.
+   *
    * @param httpHeaders Http headers
    * @param tableName Name of the table to be accessed
    * @return Whether the client has data access to the table
    */
   boolean hasDataAccess(HttpHeaders httpHeaders, String tableName);
+
+  /**
+   * Return whether the client has permission to the given table
+   *
+   * @param accessType access type
+   * @param httpHeaders HTTP headers
+   * @param tableName Name of the table to be accessed
+   * @return whether the client has permission
+   */
+  default boolean hasAccess(AccessType accessType, HttpHeaders httpHeaders, String tableName) {
+    return true;
+  }
+
+  /**
+   * Return whether the client has permission to access the endpoints which are not table level
```

Comment:
It may be good to keep this API extensible, is to add the endpoint name (and version) and the operation (CRUD) in the API. The authenticator can then choose to (for example) allow adding a new table, but not modifying any existing table or schema. Or, allow adding segments to a table, but not deleting segments.

Reply:
Sounds reasonable. Also adding the endpoint name will probably be helpful as it can be logged in AccessControl implementation.
@kishoreg since you originally proposed to use AccessType, what do you think of Subbu's suggestion?

Reply:
You may want to consider adding the endpoint URI instead of name/version.